### PR TITLE
fix(config): keep the enterprise section when merging configurations

### PR DESCRIPTION
**Pull request by jdrouet**
This PR ensures that loading the enterprise section in a different file doesn't get lost when loading those files in the good order.

Fixes https://github.com/vectordotdev/vector/issues/13290
Closes https://datadoghq.atlassian.net/browse/OP-531

Signed-off-by: Jeremie Drouet <jeremie.drouet@datadoghq.com>

### CODE REVIEW

File: src/config/builder.rs
Comment (001wwang) [inline src/config/builder.rs:212]: I think we should do something more akin to [api.merge](https://github.com/vectordotdev/vector/blob/master/src/config/api.rs#L44) than a simple override. The benefits being:
- Allows us to combine fields where it makes sense (e.g. `tags`) and error where it doesn't (e.g. `site`). Thus, keeping the user in the loop and forcing them to make conscious decisions. This should help avoid accidents. 
- Allows us to be more transparent and upfront with merging behavior with users. In the current scheme, the order in which configurations are combined matters in a non-obvious way and no indication is given that one enterprise config overrides another entirely.

**Review by 001wwang — CHANGES_REQUESTED** [review]

File: src/config/builder.rs
Comment (leebenson) [reply src/config/builder.rs:212]: Agreed. A merge strategy is in keeping with other sections. It also allows for flexible code organisation; a company could choose to keep an API key in a protected file, with remaining settings configurable by other teams in a separate file.

Comment (jdrouet) [reply src/config/builder.rs:212]: @leebenson what should happen if we merge those 2?

```toml
# first.toml
[enterprise]
enabled = false
api_key = "foo"
# second.toml
[enterprise]
enabled = true
api_key = "bar"
```

Or what if we merge those 2 considering we are not sure of the order?
```toml
# first.toml
[enterprise]
api_key = "foo"
# second.toml
[enterprise]
api_key = "bar"
```

Comment (leebenson) [reply src/config/builder.rs:212]: In both examples, the 2nd file would overwrite existing keys.

Files should be parsed alphabetically, so there's some expectation of ordering.

File: src/config/enterprise.rs
Comment (001wwang) [inline src/config/enterprise.rs:110]: I feel like we should check the contents of `site`, `region`, `endpoint` and error (or at least leave a `warn!` log) if they're mismatched when trying to merge. If a user set specific options for these across configurations, it seems accidental for them to start Vector with both (or more) disparate options.

File: src/config/enterprise.rs
Comment (001wwang) [inline src/config/enterprise.rs:120]: Similarly to above, it feels like we could at least log here for transparency if the values differ.

File: src/config/enterprise.rs
Comment (001wwang) [inline src/config/enterprise.rs:135]: We could potentially do some additional checks here for duplicate keys, leaving a warning about anything that's overwritten.

**Review by 001wwang — COMMENTED** [review]
I think this is moving in the right direction!

Before I leave more feedback, I feel like we should have a better sense of the expected UX here. In my mind, the goal is to leave the user without surprises; they should be fully aware of the final configuration they're running.

Some questions (at least for my own clarification):
- Is there precedence for simply overriding most fields in other sections (i.e. do users expect this)?
- Do we want to make important decisions for the user (i.e. which API key to select if there are many)? If so, how do we communicate said behavior to the user transparently?
- Are Vector users generally aware that configurations are combined in a specific order? (If the order is alphabetical, requiring users to tailor their configuration names for the sake of merge behavior seems like a poor UX in general.)

In my opinion, we should be cautious and error more often than not to force the user to make conscious decisions when merges conflict. At the least, we should log an `info`/`warn` whenever we overwrite.

cc: @jszwedko for any thoughts

**Comment by jszwedko** [comment]
Yeah, I agree with Will that we may want to error here when merging `[enterprise]` configurations where names collide as this is unlikely to be what the user wants.

**Comment by jdrouet** [comment]
After discussing it, we should go for a logic like 
```
if self.enterprise.is_some() && other.is_some() {
  emit error
} else {
  set self is none and other is some
}
```
instead.

**Comment by leebenson** [comment]
> After discussing it, we should go for a logic like
> 
> ```
> if self.enterprise.is_some() && other.enterprise.is_some() {
>   emit error
> } else {
>   set self is none and other is some
> }
> ```
> 
> instead.

Other merging strategies work on a field level. I think we could do the same here instead of operating against a complete `[enterprise]` block.

For example, this would be an error:

**1.toml**
```toml
[enterprise]
api_key = "xxx"
site = "bla.com"
```

**2.toml**
```toml
[enterprise]
api_key = "yyy"
site = "bla2.com"
```

... because they contain duplicate keys.

But:

**1.toml**
```toml
[enterprise]
api_key = "xxx"
```

**2.toml**
```toml
[enterprise]
site = "bla.com"
```

... could be considered valid, because the keys don't conflict, and they make up a complete enterprise block.

I'm still not convinced conflicting keys need to be anything more than a `warn!`. Like namespacing, there's precedence for requiring files be sorted alphabetically. A valid use case could be having 'root' configuration and an overrides file that contains deployment-specific exceptions. That feels like a pretty typical pattern.

I'd defer to @jszwedko on this, since this is core UX.

**Comment by 001wwang** [comment]
> Like namespacing, there's precedence for requiring files be sorted alphabetically. A valid use case could be having 'root' configuration and an overrides file that contains deployment-specific exceptions. That feels like a pretty typical pattern.

That's an interesting point, though it seems like that'd be better solved by having one template `[enterprise]` config and using environment variables? I might be misunderstanding.

In any case, the current state of the PR looks good to me: better than what we currently have and open to iteration. I agree with some points brought up earlier that we can wait until users request more advanced merging before going for it.

**Comment by jszwedko** [comment]
I'm not even sure if Vector does sort files when loading them. I think using the glob syntax does by virtue of the shell, but if I provide `--config foo.toml --config bar.toml` I'd guess `bar.toml` is _probably_ loaded after `foo.toml`. In either event, I think we can start conservative and open it up as we have motivating use-cases so I'm a fan of this going in as-is.

**Review by leebenson — APPROVED** [review]
Thanks everyone for the discussion 👍🏻 

Makes sense to start restrictive and open it up later if requested.

LGTM, subject to fixing the merge conflict.

**Review by 001wwang — APPROVED** [review]
Thanks for iterating on this! Merging so this can make it into the `0.23.0` release.

**Pull request merged**
